# 6.1) U.S.A. REAL ESTATE 

## -01 import libraries & data
## -02 data checks
     df.shape, df.head(), df.tail(), df.info(), df.describe(), df.columns, 
###      a) check mixed data types (change 3 columns to string)
###      b) check missing values 
     create subsets and impute means/medians as appropriate to their columns
###      c) find and address duplicates
## -03 export new data, as re_clean.csv     
     

# 01 import libraries & data

#import libraries
import pandas as pd
import numpy as np
import os


In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#import path
path = r'/Users/nancykray/Desktop/Real Estate Project'

In [3]:
path

'/Users/nancykray/Desktop/Real Estate Project'

In [4]:
#import data (csv)
df=pd.read_csv(os.path.join(path,'Original Data','realtor-data.zip.csv'), index_col=False)

# 02 data checks

In [5]:
df.shape

(2226382, 12)

In [6]:
df.head()

brokered_by    status     price  bed  bath  acre_lot     street  \
0     103378.0  for_sale  105000.0  3.0   2.0      0.12  1962661.0   
1      52707.0  for_sale   80000.0  4.0   2.0      0.08  1902874.0   
2     103379.0  for_sale   67000.0  2.0   1.0      0.15  1404990.0   
3      31239.0  for_sale  145000.0  4.0   2.0      0.10  1947675.0   
4      34632.0  for_sale   65000.0  6.0   2.0      0.05   331151.0   

         city        state  zip_code  house_size prev_sold_date  
0    Adjuntas  Puerto Rico     601.0       920.0            NaN  
1    Adjuntas  Puerto Rico     601.0      1527.0            NaN  
2  Juana Diaz  Puerto Rico     795.0       748.0            NaN  
3       Ponce  Puerto Rico     731.0      1800.0            NaN  
4    Mayaguez  Puerto Rico     680.0         NaN            NaN

In [7]:
#checking on dates to make sure at least in past 3 years since there were NaNs in the first view using head()
df.tail()

brokered_by status     price  bed  bath  acre_lot     street  \
2226377      23009.0   sold  359900.0  4.0   2.0      0.33   353094.0   
2226378      18208.0   sold  350000.0  3.0   2.0      0.10  1062149.0   
2226379      76856.0   sold  440000.0  6.0   3.0      0.50   405677.0   
2226380      53618.0   sold  179900.0  2.0   1.0      0.09   761379.0   
2226381     108243.0   sold  580000.0  5.0   3.0      0.31   307704.0   

             city       state  zip_code  house_size prev_sold_date  
2226377  Richland  Washington   99354.0      3600.0     2022-03-25  
2226378  Richland  Washington   99354.0      1616.0     2022-03-25  
2226379  Richland  Washington   99354.0      3200.0     2022-03-24  
2226380  Richland  Washington   99354.0       933.0     2022-03-24  
2226381  Richland  Washington   99354.0      3615.0     2022-03-23

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2226382 entries, 0 to 2226381
Data columns (total 12 columns):
 #   Column          Dtype  
---  ------          -----  
 0   brokered_by     float64
 1   status          object 
 2   price           float64
 3   bed             float64
 4   bath            float64
 5   acre_lot        float64
 6   street          float64
 7   city            object 
 8   state           object 
 9   zip_code        float64
 10  house_size      float64
 11  prev_sold_date  object 
dtypes: float64(8), object(4)
memory usage: 203.8+ MB


In [9]:
#check descriptive statistics
df.describe()

brokered_by         price           bed          bath      acre_lot  \
count  2.221849e+06  2.224841e+06  1.745065e+06  1.714611e+06  1.900793e+06   
mean   5.293989e+04  5.241955e+05  3.275841e+00  2.496440e+00  1.522303e+01   
std    3.064275e+04  2.138893e+06  1.567274e+00  1.652573e+00  7.628238e+02   
min    0.000000e+00  0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    2.386100e+04  1.650000e+05  3.000000e+00  2.000000e+00  1.500000e-01   
50%    5.288400e+04  3.250000e+05  3.000000e+00  2.000000e+00  2.600000e-01   
75%    7.918300e+04  5.500000e+05  4.000000e+00  3.000000e+00  9.800000e-01   
max    1.101420e+05  2.147484e+09  4.730000e+02  8.300000e+02  1.000000e+05   

             street      zip_code    house_size  
count  2.215516e+06  2.226083e+06  1.657898e+06  
mean   1.012325e+06  5.218668e+04  2.714471e+03  
std    5.837635e+05  2.895408e+04  8.081635e+05  
min    0.000000e+00  0.000000e+00  4.000000e+00  
25%    5.063128e+05  2.961700e+04  1.300000e+03  
50%    1.012766e+06  4.838200e+04  1.760000e+03  
75%    1.521173e+06  7.807000e+04  2.413000e+03  
max    2.001357e+06  9.999900e+04  1.040400e+09

In [10]:
# Check the 12 columns that the describe() function referred to
df.columns

Index(['brokered_by', 'status', 'price', 'bed', 'bath', 'acre_lot', 'street',
       'city', 'state', 'zip_code', 'house_size', 'prev_sold_date'],
      dtype='object')

In [11]:
#using df.value_counts() function to find missing values/Nans within a column
df['brokered_by'].value_counts(dropna=False)

brokered_by
22611.0    45658
16829.0    27732
53016.0    21709
23592.0     9176
30807.0     8464
           ...  
36774.0        1
88410.0        1
92528.0        1
76894.0        1
98730.0        1
Name: count, Length: 110144, dtype: int64

## 02 a) check mixed data types

In [12]:
#find mixed values
for col in df.columns.tolist():
    weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df[weird])>0:
        print(col)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2090348899.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2090348899.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2090348899.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2090348899.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fb/2jrdrrb5

city
state


/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2090348899.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2090348899.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2090348899.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)


prev_sold_date


### 3 columns identified as mixed data type, now I will change to string

In [13]:
#changing the 3 columns (city, state, prev_sold_date) to string
df['city'] = df['city'].astype('str')
df['state'] = df['state'].astype('str')
df['prev_sold_date'] = df['prev_sold_date'].astype('str')

In [14]:
#check dtypes again now that changes are made
df.dtypes

brokered_by       float64
status             object
price             float64
bed               float64
bath              float64
acre_lot          float64
street            float64
city               object
state              object
zip_code          float64
house_size        float64
prev_sold_date     object
dtype: object

In [15]:
#re-check for mixed data types again (after changes)
for col in df.columns.tolist():
    weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df[weird])>0:
        print(col)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2931238424.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2931238424.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2931238424.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/2931238424.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type)!=df[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fb/2jrdrrb5

#### ok, looks good now

## 02 b) check missing values

In [16]:
#to find the Nans in every column
df.isnull().sum()

brokered_by         4533
status                 0
price               1541
bed               481317
bath              511771
acre_lot          325589
street             10866
city                   0
state                  0
zip_code             299
house_size        568484
prev_sold_date         0
dtype: int64

### 8 rows have several missing values
#### 'brokered_by' is ID numbers so imputing doesn't make sense - i'll have to remove into subset and put somewhere:  since there are 4533 NaNs, the new dataset will account for these removed: 2221849

### create subsets of Nans for columns with missing values

In [17]:
#create a subset of only the missing values
df_nan_brokered_by = df[df['brokered_by'].isnull()==True]

In [18]:
df_nan_brokered_by

brokered_by    status     price  bed  bath  acre_lot     street  \
518              NaN  for_sale   55000.0  NaN   NaN      0.85   599860.0   
990              NaN  for_sale  255000.0  3.0   2.0       NaN   595720.0   
1476             NaN  for_sale  135000.0  3.0   1.0      0.11  1450443.0   
10053            NaN  for_sale  679900.0  6.0   4.0      1.82     7650.0   
13511            NaN  for_sale  309900.0  2.0   1.0      0.11   843765.0   
...              ...       ...       ...  ...   ...       ...        ...   
2220033          NaN      sold  435000.0  3.0   2.0      1.10   817222.0   
2221167          NaN      sold  375000.0  3.0   2.0      1.39  1692463.0   
2225547          NaN      sold  245000.0  2.0   2.0      0.13   576308.0   
2225844          NaN      sold  358000.0  3.0   2.0      0.26  1372469.0   
2225864          NaN      sold  409000.0  4.0   2.0      0.19   480205.0   

                     city         state  zip_code  house_size prev_sold_date  
518           Hormigueros   Puerto Rico     660.0         NaN            nan  
990              Guaynabo   Puerto Rico     965.0      1500.0            nan  
1476              Bayamon   Puerto Rico     957.0       782.0            nan  
10053            Cromwell   Connecticut    6416.0      3184.0     2011-06-08  
13511    North Providence  Rhode Island    2904.0      1016.0     1992-11-17  
...                   ...           ...       ...         ...            ...  
2220033    East Wenatchee    Washington   98802.0      1512.0     2022-01-12  
2221167        Waterville    Washington   98858.0      1512.0     2022-02-15  
2225547             Pasco    Washington   99301.0      1500.0     2022-04-12  
2225844     West Richland    Washington   99353.0      1582.0     2022-04-13  
2225864          Richland    Washington   99352.0      2121.0     2021-12-03  

[4533 rows x 12 columns]

In [19]:
#create a new dataset with these NaNs removed from 'brokered_by' column
df_clean = df[df['brokered_by'].isnull() == False]

In [20]:
#compare new shape
df_clean.shape

(2221849, 12)

### good, this is the expected amount (2226382-4533)

### decide whether to impute mean or median for missing values

#### checking median for appropriate variables (choose median if outliers pull mean up/down)

In [21]:
df_clean[['brokered_by', 'price', 'bed', 'bath', 'acre_lot', 'street','zip_code', 'house_size']].median()

brokered_by      52884.00
price           325000.00
bed                  3.00
bath                 2.00
acre_lot             0.26
street         1012986.50
zip_code         48433.00
house_size        1760.00
dtype: float64

#### checking mean again to compare

In [22]:
df_clean.describe()

brokered_by         price           bed          bath      acre_lot  \
count  2.221849e+06  2.220308e+06  1.742245e+06  1.711817e+06  1.896744e+06   
mean   5.293989e+04  5.240079e+05  3.275919e+00  2.496373e+00  1.524812e+01   
std    3.064275e+04  2.135990e+06  1.567769e+00  1.653039e+00  7.636356e+02   
min    0.000000e+00  0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    2.386100e+04  1.650000e+05  3.000000e+00  2.000000e+00  1.500000e-01   
50%    5.288400e+04  3.250000e+05  3.000000e+00  2.000000e+00  2.600000e-01   
75%    7.918300e+04  5.500000e+05  4.000000e+00  3.000000e+00  9.900000e-01   
max    1.101420e+05  2.147484e+09  4.730000e+02  8.300000e+02  1.000000e+05   

             street      zip_code    house_size  
count  2.211000e+06  2.221550e+06  1.655219e+06  
mean   1.012480e+06  5.219751e+04  2.715133e+03  
std    5.838324e+05  2.895915e+04  8.088170e+05  
min    0.000000e+00  0.000000e+00  4.000000e+00  
25%    5.063548e+05  2.961500e+04  1.300000e+03  
50%    1.012986e+06  4.843300e+04  1.760000e+03  
75%    1.521461e+06  7.807100e+04  2.413000e+03  
max    2.001357e+06  9.999900e+04  1.040400e+09

In [44]:
#index/create subset to check the outliers of the max for price (to figure out to impute median or mean)
df_high_price=df_clean[df_clean['price']>2000000000]

In [45]:
df_high_price

brokered_by    status         price  bed  bath  acre_lot   street  \
221994      23826.0  for_sale  2.147484e+09  2.0   2.0      0.12  11355.0   

                 city       state  zip_code  house_size prev_sold_date  
221994  International  California       NaN       885.0            nan

### there is only one house sold for this extremely high amount.  When I deleted a 0, there was only a handful in those high prices as well.  Therefore, I will impute the median in these NaNs (instead of mean).

In [57]:
df_big_acre

brokered_by    status       price   bed  bath   acre_lot     street  \
3303         88715.0  for_sale    585000.0   4.0   4.0  100000.00  1930360.0   
5263         10554.0  for_sale    150000.0   NaN   NaN   99999.00   595393.0   
9189         10560.0  for_sale   5700000.0   NaN   NaN   99999.00  1106245.0   
22315        76006.0  for_sale    250000.0   3.0   3.0   18817.00  1670131.0   
26954        42180.0  for_sale    319000.0   3.0   1.0   13697.00  1407338.0   
...              ...       ...         ...   ...   ...        ...        ...   
2121326      99588.0      sold    685000.0  12.0   8.0   10454.00  1035173.0   
2123142      53081.0      sold    399000.0   3.0   2.0   39077.00  1250202.0   
2123611      65227.0      sold    589000.0   NaN   NaN   14733.00   669875.0   
2175102      65337.0      sold    138000.0   NaN   NaN   43996.00   547688.0   
2190442      18836.0      sold  15550000.0   2.0   1.0   28274.29  1227775.0   

                 city         state  zip_code  house_size prev_sold_date  
3303         San Juan   Puerto Rico     926.0      3300.0            nan  
5263         Simsbury   Connecticut    6070.0         NaN            nan  
9189       Middletown   Connecticut    6457.0     99999.0            nan  
22315    Narragansett  Rhode Island    2882.0      1532.0     2020-09-14  
26954        Portland         Maine    4108.0       992.0            nan  
...               ...           ...       ...         ...            ...  
2121326   Bakersfield    California   93313.0      4597.0     2021-11-23  
2123142   Quartz Hill    California   93536.0      1239.0     2022-04-12  
2123611     Tehachapi    California   93561.0      3360.0     2022-04-25  
2175102  Grass Valley    California   95949.0         NaN     2022-04-01  
2190442       Ashwood        Oregon   97711.0      2330.0     2021-12-22  

[361 rows x 12 columns]

### only small amount (361) with such big acre_lot, I will use median to impute as well (instead of mean)

## impute  median for 'price', 'zip', 'acre_lot', & 'street'.  For all others impute mean?  

In [60]:
#find median first
df_clean['price'].median()

325000.0

In [62]:
#impute mean price in all NaN values 
df_clean['price'].fillna(325000.0, inplace=True)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/1482927526.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['price'].fillna(325000.0, inplace=True)


In [65]:
#find median of zip
df_clean['zip_code'].median()

48433.0

In [67]:
#impute median zip_code in all NaN values
df_clean['zip_code'].fillna(48433.0, inplace=True)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/303787387.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['zip_code'].fillna(48433.0, inplace=True)


In [69]:
#find median for acre_lot
df_clean['acre_lot'].median()

0.26

In [70]:
#impute median of acre_lot in all NaN values
df_clean['acre_lot'].fillna(.26, inplace=True)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/1778491911.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['acre_lot'].fillna(.26, inplace=True)


In [74]:
#find median of street
df_clean['street'].median()

1012986.5

In [75]:
#impute median of street in all NaN values
df_clean['street'].fillna(1012986.5, inplace=True)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/31305090.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['street'].fillna(1012986.5, inplace=True)


## impute mean now for all Nans in 'bed', 'bath', & 'house_size'

In [71]:
#impute mean for bed
df_clean['bed'].fillna(3.275919, inplace=True)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/1568750744.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['bed'].fillna(3.275919, inplace=True)


In [72]:
#impute mean for bath
df_clean['bath'].fillna(2.496373, inplace=True)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/3446227760.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['bath'].fillna(2.496373, inplace=True)


In [73]:
#impute mean for house_size
df_clean['house_size'].fillna(2.715133e+03 , inplace=True)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_20120/1133291821.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['house_size'].fillna(2.715133e+03 , inplace=True)


## Re-check NaNs in df_clean dataset to make sure imputed correctly 

In [76]:
df_clean.isnull().sum()

brokered_by       0
status            0
price             0
bed               0
bath              0
acre_lot          0
street            0
city              0
state             0
zip_code          0
house_size        0
prev_sold_date    0
dtype: int64

## looks good since only 0s have been returned - meaning no missing values found

## 02 c) Find and Address Duplicates

In [77]:
#find full duplicates in dataframe (repeated across entire row)
df_dups = df_clean[df_clean.duplicated()]

In [78]:
df_dups

brokered_by    status     price       bed      bath  acre_lot  \
978216      96698.0  for_sale   61900.0  3.275919  2.496373      0.26   
978245      96698.0  for_sale  144900.0  3.275919  2.496373      0.26   
981925      54857.0  for_sale   26500.0  3.275919  2.496373      0.26   

           street         city     state  zip_code  house_size prev_sold_date  
978216  1840867.0  Hot Springs  Arkansas   71901.0    2715.133            nan  
978245  1840866.0  Hot Springs  Arkansas   71901.0    2715.133            nan  
981925  1012986.5       Lonoke  Arkansas   72086.0    2715.133            nan

In [79]:
#check the shape prior to removing duplicates
df_clean.shape

(2221849, 12)

In [80]:
#create a dataframe first that doesn't include duplicates
df_clean_no_dups = df_clean.drop_duplicates()

In [82]:
#check shape of this new dataframe to compare above (should have 3 less to account for 3 duplicates removed)
df_clean_no_dups.shape

(2221846, 12)

# 03 Export cleaned dataset

In [84]:
df_clean_no_dups.to_csv(os.path.join(path,'Prepared Data', 're_clean.csv'))